uxarray/grid/neighbors.py

```diff
@@ -3,11 +3,236 @@
+
+        return _tree
+
+    def query(self,
+              xyz: Union[np.ndarray, list, tuple],
+              k: Optional[int] = 1,
+              return_distance: Optional[bool] = True,
+              dualtree: Optional[bool] = False,
+              breadth_first: Optional[bool] = False,
+              sort_results: Optional[bool] = True):
+        """Queries the tree for the ``k`` nearest neighbors.
+
+        Parameters
+        ----------
+        xyz : array_like
+            coordinate pairs in cartesian (x, y, z) to query
+        k: int, default=1
+            The number of nearest neighbors to return
+        return_distance : bool, optional
+            Indicates whether distances should be returned
+        dualtree : bool, default=False
+            Indicates whether to use the dual-tree formalism for node queries
+        breadth_first : bool, default=False
+            Indicates whether to query nodes in a breadth-first manner
+        sort_results : bool, default=True
+            Indicates whether distances should be sorted
+
+        Returns
+        -------
+        d : ndarray of shape (xyz.shape[0], k), dtype=double
+            Distance array that keeps the distances of the k-nearest neighbors to the entries from xyz in each row
+        ind : ndarray of shape (xyz.shape[0], k), dtype=INT_DTYPE
+            Index array that keeps the indices of the k-nearest neighbors to the entries from xyz in each row
+        """
+
+        if k < 1 or k > self._n_elements:
+            raise AssertionError(
+                f"The value of k must be greater than 1 and less than the number of elements used to construct "
+                f"the tree ({self._n_elements}).")
+
+        xyz = _prepare_cart_xyz_for_query(xyz)
+
+        d, ind = self._current_tree().query(xyz, k, return_distance, dualtree,
+                                            breadth_first, sort_results)
+
+        ind = np.asarray(ind, dtype=INT_DTYPE)
+
+        if xyz.shape[0] == 1:
+            ind = ind.squeeze()
+
+        # perform query with distance
+        if return_distance:
+            # only one pair was queried
+            if xyz.shape[0] == 1:
+                d = d.squeeze()
+
+            return d, ind
+
+        return ind
+
+    def query_radius(self,
+                     xyz: Union[np.ndarray, list, tuple],
+                     r: Optional[int] = 1.0,
+                     return_distance: Optional[bool] = True,
+                     count_only: Optional[bool] = False,
+                     sort_results: Optional[bool] = False):
+        """Queries the tree for all neighbors within a radius ``r``.
+
+        Parameters
+        ----------
+        xyz : array_like
+           coordinate pairs in cartesian (x, y, z) to query
+        r: distance in degrees within which neighbors are returned
+            r can be a single value , or an array of values of shape x.shape[:-1] if different radii are desired for each point.
```

Comment:
@philipc2 I haven't actually seen the documentation for having multiple radii, are we sure this is something that works? I sent in an array of points and a matching array of radii but it didn't accept that input.

Reply:
Good catch, I think I may have copied this docstring section for `sklearn` directly without thinking about it. I'm not sure if it works. I think it would be OK to remove that part from both the KD and Ball Tree and we can investigate it in the future.